### PR TITLE
Debian: Added support for versions 10, 11 and 12

### LINKED ISSUE #24: Debian 10 template
Hello Team, 

I would like to know if we will have a template for Debian 10.

Thanks.

### PR DESCRIPTION
**Pull request by alanbach**
Added experimental support for Debian 10 (Buster), 11 (Bullseye) & 12 (Bookworm).

Closes #24

### CODE REVIEW

**Review by alexsander-souza — CHANGES_REQUESTED** [review]
awesome contribution, but this needs some cleanup

- remove changes to the `ubuntu` template, I believe they belong to another PR
- remove unused scripts from `debian/scripts`, I think they were copied from `ubuntu` but not required by the image you are building.

**Comment by alanbach** [comment]
> awesome contribution, but this needs some cleanup
> 
>     * remove changes to the `ubuntu` template, I believe they belong to another PR
> 
>     * remove unused scripts from `debian/scripts`, I think they were copied from `ubuntu` but not required by the image you are building.

These changes have been applied.

**Comment by proegssilb** [comment]
@alanbach Did you changes make it into the PR? Github is claiming that you still have changes to Ubuntu in the PR.

**Comment by alanbach** [comment]
> @alanbach Did your changes make it into the PR? Github is claiming that you still have changes to Ubuntu in the PR.

@proegssilb  I don't know why changes that were a part of a different PR are showing up here. Those are already merged and this looks to me some sort of a snag.

**Comment by proegssilb** [comment]
Looks like that merge cleaned up the Ubuntu changes. Didn't check the scripts, but hopefully this PR can be re-reviewed soon. Thanks alanbach!

**Comment by alanbach** [comment]
Yup, the branch re-sync cleaned it up. Give this a try and let me know if you hit any issues! Hopefully we can get this merged soon.

**Comment by proegssilb** [comment]
Still working on troubleshooting my setup. I'll post if there's any issues, but I wouldn't advise using my difficulties as an excuse to delay this PR. I probably just need to read through the README and logs in more detail until I can figure out what's going on. And then maybe we can update the README in a future PR.

**Comment by erickellerek1** [comment]
also trying to use this template to build debian arm64 maas images.
I found out that the readme is missing some dependencies for arm: `qemu-system-arm qemu-efi-aarch64` maybe adding these would prevent people from getting following qemu/packer error: `cannot stat '/usr/share/AAVMF/AAVMF_VARS.fd': No such file or directory`

File: debian/scripts/cloudimg/setup-boot.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+if  [ ! -z  "${https_proxy}" ]; then
+  echo "Acquire::https::Proxy \"${https_proxy}\";" >> ${packer_apt_proxy_config}
+fi
+
+# Reset cloud-init, so that it can run again when MAAS deploy the image.
+cloud-init clean --logs
+
+apt-get update
+if [ ${BOOT_MODE} == "uefi" ] && [ ${DEBIAN_VERSION} == '12' ]; then
+        apt-get install -y grub-cloud-amd64 grub-efi-amd64
```
Comment (erickellerek1) [inline debian/scripts/cloudimg/setup-boot.sh:36]: we shall parametrize the architecture as installing amd64 package on arm64 will obviously not work and get following package manager errors:
```
==> cloudimg.image.qemu.cloudimg: E: Package 'grub-cloud-amd64' has no installation candidate
==> cloudimg.image.qemu.cloudimg: E: Package 'grub-efi-amd64' has no installation candidate
```

File: debian/scripts/cloudimg/setup-boot.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+else
+        apt-get install -y grub-cloud-amd64 grub-pc
+fi
+
+# Bookworm does not include this, but curtin requires this during the installation.
+if [ ${DEBIAN_VERSION} == '12' ]; then
+        wget http://ftp.us.debian.org/debian/pool/main/e/efibootmgr/efibootmgr_15-1_amd64.deb
```
Comment (erickellerek1) [inline debian/scripts/cloudimg/setup-boot.sh:43]: same would apply here, for ARM architecture we cannot expect this to work.

File: debian/scripts/setup-bootloader [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+fi
+
+# Clean up remnants from packer-maas vm install
+rm /var/cache/debconf/config.dat
+dpkg --configure -a
+
+apt-get update
+
+if [ -f /sys/firmware/efi/runtime ]; then
+    apt-get install -y grub-cloud-amd64 grub-efi-amd64
```
Comment (erickellerek1) [inline debian/scripts/setup-bootloader:40]: same architecture comments as before ARM architecture target is not going to accept amd64 architecture.

**Review by erickellerek1 — COMMENTED** [review]
using this template for ARM64 architecture would require some minor changes.

**Comment by lwandrebeck** [comment]
MAAS snap latest/edge (thus 3.5.0-alpha1), with packer-maas on your PR-24 branch.
flat storage, / ext4, /var xfs.
Image built with `sudo packer build -var kernel=linux-image-amd64 -var
customize_script=tweaks.sh -var debian_series=bookworm -var
debian_version=12 -only='cloudimg.*' .`
tweaks.sh contains only `apt-get install -y lvm2 xfsprogs` otherwise
curtin fails with `E: Unable to locate package lvm2` or `E: Unable to
locate package xfsprogs`
Image uploaded with `sudo maas $LOGIN boot-resources create
name='custom/debian-12' title='debian-12' architecture='amd64/generic'
filetype='tgz' content@=debian-custom-cloudimg.tar.gz`
`/var/snap/maas/current/preseeds/curtin_userdata_custom` is a copy of
the file from this PR.
machine boots, BUT once arrived at systemd I get:
- hostname not properly defined
- systemd is slow (30 sec per step)
- network looks broken




Trying to dig why.
`

**Comment by lwandrebeck** [comment]
btw, /var/snap/maas/current/preseeds/curtin_userdata_custom breaks known to work tweaked ubuntu image deployment, works again with file removed. Don’t know MAAS internals enough to know if a conditional (on uname or something ?) would do the trick to prevent other OSes breakage.

**Comment by alexsander-souza** [comment]
> btw, /var/snap/maas/current/preseeds/curtin_userdata_custom breaks known to work tweaked ubuntu image deployment, works again with file removed. Don’t know MAAS internals enough to know if a conditional (on uname or something ?) would do the trick to prevent other OSes breakage.

you can name the preseed as `curtin_userdata_custom_amd64_generic_debian-10`, assuming `amd64/generic` as node architecture and that you called the image `custom/debian-10` when uploading

File: debian/README.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+The default username is ```debian```
+
+## Uploading images to MAAS
+
+TGZ image
+
+```shell
+maas admin boot-resources create \
+    name='custom/debian-tgz' \
```
Comment (alexsander-souza) [inline debian/README.md:130]: change to `custom/debian-12` or something like this

File: debian/scripts/curtin-hooks
Comment (alexsander-souza) [inline debian/scripts/curtin-hooks:1]: not used

File: debian/scripts/cloudimg/cleanup.sh
Comment (alexsander-souza) [inline debian/scripts/cloudimg/cleanup.sh:1]: move to parent directory

File: debian/scripts/cloudimg/curtin-hooks
Comment (alexsander-souza) [inline debian/scripts/cloudimg/curtin-hooks:1]: move to parent directory

File: debian/scripts/cloudimg/install-custom-kernel.sh
Comment (alexsander-souza) [inline debian/scripts/cloudimg/install-custom-kernel.sh:1]: move to parent directory

File: debian/scripts/cloudimg/setup-boot.sh
Comment (alexsander-souza) [inline debian/scripts/cloudimg/setup-boot.sh:1]: move to parent directory

File: debian/scripts/cloudimg/setup-curtin.sh
Comment (alexsander-souza) [inline debian/scripts/cloudimg/setup-curtin.sh:1]: move to parent directory

**Comment by proegssilb** [comment]
> you can name the preseed as `curtin_userdata_custom_amd64_generic_debian-10`, assuming `amd64/generic` as node architecture and that you called the image `custom/debian-10` when uploading

Can we get this into the readme, or link to the relevant documentation from the readme?

File: debian/packages/get-kernel.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+export LANG=C
+
+KERNEL=${KERNEL:-linux-image-generic}
+ARCH=amd64
```
Comment (alexsander-souza) [inline debian/packages/get-kernel.sh:24]: call `dpkg --print-architecture` instead

**Comment by lwandrebeck** [comment]
> > btw, /var/snap/maas/current/preseeds/curtin_userdata_custom breaks known to work tweaked ubuntu image deployment, works again with file removed. Don’t know MAAS internals enough to know if a conditional (on uname or something ?) would do the trick to prevent other OSes breakage.
> 
> you can name the preseed as `curtin_userdata_custom_amd64_generic_debian-10`, assuming `amd64/generic` as node architecture and that you called the image `custom/debian-10` when uploading

works fine ! Thanks for the tip @alexsander-souza :+1:

File: debian/scripts/setup-bootloader (hunk 3 above)
Comment (alanbach) [reply debian/scripts/setup-bootloader:40]: Agreed & thank you for these. I'll work on a proper arm64 support and will incorporate these.

**Comment by alanbach** [comment]
> MAAS snap latest/edge (thus 3.5.0-alpha1), with packer-maas on your PR-24 branch. flat storage, / ext4, /var xfs. Image built with `sudo packer build -var kernel=linux-image-amd64 -var customize_script=tweaks.sh -var debian_series=bookworm -var debian_version=12 -only='cloudimg.*' .` tweaks.sh contains only `apt-get install -y lvm2 xfsprogs` otherwise curtin fails with `E: Unable to locate package lvm2` or `E: Unable to locate package xfsprogs` Image uploaded with `sudo maas $LOGIN boot-resources create name='custom/debian-12' title='debian-12' architecture='amd64/generic' filetype='tgz' content@=debian-custom-cloudimg.tar.gz` `/var/snap/maas/current/preseeds/curtin_userdata_custom` is a copy of the file from this PR. machine boots, BUT once arrived at systemd I get:

Good catch with lvm/xfsprogs. I'll get that fixed.

File: debian/README.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
+# Debian Packer Templates for MAAS
+
+## Introduction
+
+The Packer templates in this directory creates Debian images for use with MAAS.
+
+## Prerequisites (to create the image)
+
+* A machine running Debian 18.04+ with the ability to run KVM virtual machines.
```
Comment (sparkiegeek) [inline debian/README.md:9]: what is this Debian 18.04+ of which you speak?

Comment (alanbach) [reply debian/README.md:9]: Fixing...

File: debian/packages/get-kernel.sh (hunk 5 above)
Comment (alanbach) [reply debian/packages/get-kernel.sh]: Removed as not being used.

File: debian/scripts/networking.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+rm -f /var/log/cloud-init*.log
+rm -rf /var/lib/cloud/instances \
+    /var/lib/cloud/instance
+
+# Install a dpkg-query wrapper to mypass MAAS netplan.io check 
+echo "IyEvYmluL3NoClsgIiQxIiA9ICctcycgXSAmJiBbICIkMiIgPSAnbmV0cGxhbi5pbycgXSAmJiBleGl0IDAKL3Vzci9iaW4vZHBrZy1xdWVyeSAiJEAiCg==" | base64 -d > /usr/local/bin/dpkg-query
+chmod 755 /usr/local/bin/dpkg-query
+
+# Debian netplan.io does not have an info parameter, work around it
+echo "IyEvYmluL3NoClsgIiQxIiA9ICdpbmZvJyBdICYmIGV4aXQgMAovdXNyL3NiaW4vbmV0cGxhbiAiJEAiCg==" | base64 -d > /usr/local/bin/netplan
```
Comment (alexsander-souza) [inline debian/scripts/networking.sh:53]:
```suggestion
cat > /usr/local/bin/netplan <<EOF
#!/bin/sh
[ "\$1" = 'info' ] && exit 0
/usr/sbin/netplan "\$@"
EOF
```

File: debian/scripts/networking.sh [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+rm -f /etc/cloud/cloud.cfg.d/subiquity-disable-cloudinit-networking.cfg
+rm -f /etc/cloud/cloud.cfg.d/99-installer.cfg
+rm -f /etc/cloud/ds-identify.cfg
+
+rm -f /var/log/cloud-init*.log
+rm -rf /var/lib/cloud/instances \
+    /var/lib/cloud/instance
+
+# Install a dpkg-query wrapper to mypass MAAS netplan.io check 
+echo "IyEvYmluL3NoClsgIiQxIiA9ICctcycgXSAmJiBbICIkMiIgPSAnbmV0cGxhbi5pbycgXSAmJiBleGl0IDAKL3Vzci9iaW4vZHBrZy1xdWVyeSAiJEAiCg==" | base64 -d > /usr/local/bin/dpkg-query
```
Comment (alexsander-souza) [inline debian/scripts/networking.sh:49]:
```suggestion
cat > /usr/local/bin/dpkg-query <<EOF
#!/bin/sh
[ "\$1" = '-s' ] && [ "\$2" = 'netplan.io' ] && exit 0
/usr/bin/dpkg-query "\$@"
EOF
```

File: debian/scripts/networking.sh [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+rm -f /etc/cloud/cloud.cfg.d/subiquity-disable-cloudinit-networking.cfg
+rm -f /etc/cloud/cloud.cfg.d/99-installer.cfg
+rm -f /etc/cloud/ds-identify.cfg
+
+rm -f /var/log/cloud-init*.log
+rm -rf /var/lib/cloud/instances \
+    /var/lib/cloud/instance
+
+# Install a dpkg-query wrapper to mypass MAAS netplan.io check
```
Comment (alexsander-souza) [inline debian/scripts/networking.sh:48]: typo? `mypass` -> `by-pass`

**Review by alexsander-souza — COMMENTED** [review]
let's expand these wrappers, otherwise it's hard to tell what the template is doing

**Comment by proegssilb** [comment]
Lightly tested commit [319fc01](https://github.com/canonical/packer-maas/pull/162/commits/319fc016a569e57231257ada729f6f6a25429aa9) . Compiling amd64 images via `make` seems to work OK. Whatever issues I was having before seem to be gone.

**Comment by alanbach** [comment]
Thank you. Currently working on completing the arm64 support.

**Comment by alanbach** [comment]
OK, just tested building and deploying arm64 Buster, Bullseye and Bookworm images. They all worked, just a small note that the preseed file needs to be updated for arm64 since the kernel package name is linux-image-arm64. 

Updated the README file and included a separate ready to use arm64 preseed file as well. 

Please try out the new changes and let me know if you see any issues.

**Review by alexsander-souza — APPROVED** [review]
LGTM

**Comment by lwandrebeck** [comment]
quick test, bookworm installed and worked fine on x86_64 with your latest commits ! thanks a lot @alanbach

**Comment by lwandrebeck** [comment]
Debian 11 worked too. 10 fails like on yesterday photos.

**Pull request merged**